pkg/virt-launcher/fork.go

```diff
@@ -0,0 +1,143 @@
+	defer cleanupContainerDiskDirectory(containerDiskDir)
+	index := 0
+	if !nonRoot {
+		index = 1
+	}
+	cmd := exec.Command(os.Args[index], append(os.Args[(index+1):], "--no-fork", "true")...)
+	if !nonRoot {
+		cmd.SysProcAttr = &syscall.SysProcAttr{
+			AmbientCaps: []uintptr{10},
+		}
+	}
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Start(); err != nil {
+		log.Log.Reason(err).Error("failed to fork virt-launcher")
+		return 1, err
+	}
+
+	exitStatus := make(chan syscall.WaitStatus, 10)
+	sigs := make(chan os.Signal, 10)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
+	go func() {
+		for sig := range sigs {
+			switch sig {
+			case syscall.SIGCHLD:
+				var wstatus syscall.WaitStatus
+				wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
+				if err != nil {
+					log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
+				}
+
+				// there's a race between cmd.Wait() and syscall.Wait4 when
+				// cleaning up the cmd's pid after it exits. This allows us
+				// to detect the correct exit code regardless of which wait
+				// wins the race.
+				if wpid == cmd.Process.Pid {
+					exitStatus <- wstatus
+				}
+
+			default:
+				log.Log.V(3).Log("signalling virt-launcher to shut down")
+				err := cmd.Process.Signal(syscall.SIGTERM)
+				sig.Signal()
+				if err != nil {
+					log.Log.Reason(err).Errorf("received signal %s but can't signal virt-launcher to shut down", sig.String())
+				}
+			}
+		}
+	}()
+
+	// wait for virt-launcher and collect the exit code
+	exitCode := 0
+	if err := cmd.Wait(); err != nil {
+		select {
+		case status := <-exitStatus:
+			exitCode = int(status)
+		default:
+			exitCode = 1
+			if exiterr, ok := err.(*exec.ExitError); ok {
+				if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+					exitCode = status.ExitStatus()
+				}
+			}
+			log.Log.Reason(err).Error("dirty virt-launcher shutdown")
+		}
+
+	}
+	// give qemu some time to shut down in case it survived virt-handler
+	// Most of the time we call `qemu-system=* binaries, but qemu-system-* packages
+	// are not everywhere available where libvirt and qemu are. There we usually call qemu-kvm
+	// which resides in /usr/libexec/qemu-kvm
+	pid, _ := FindPid("qemu-system")
+	qemuProcessCommandPrefix := "qemu-system"
+	if pid <= 0 {
+		pid, _ = FindPid("qemu-kvm")
+		qemuProcessCommandPrefix = "qemu-kvm"
+	}
+	if pid > 0 {
+		p, err := os.FindProcess(pid)
+		if err != nil {
+			return 1, err
+		}
+		// Signal qemu to shutdown
+		err = p.Signal(syscall.SIGTERM)
+		if err != nil {
+			return 1, err
+		}
+		// Wait for 10 seconds for the qemu process to disappear
+		err = utilwait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+			pid, _ := FindPid(qemuProcessCommandPrefix)
+			if pid == 0 {
+				return true, nil
+			}
+			return false, nil
+		})
+		if err != nil {
+			return 1, err
+		}
+	}
```

Comment:
I am not sure why this code block is needed for the 'virt-launcher-cap' binary. 

Can you please discuss a little the reason for having the 'forkandmonitor' process in the 'virt-launcher-cap'?

Reply:
Currently, we launch virt-launcher and it will fork itself. To have the same behavior we now launch the virt-launcher-cap and it will fork the virt-launcher binary in case of non-root(root will do the same as before).

Reply:
I saw you have had some 'commented-out' code in non-root.go which appears sufficient to me. 

For now, in the forkandmonitor, when being called from non-root.go, to me, there are a large number of blocks of code-path will or should not be executed. For one example, the block that brings-down the qemu should not be executed. 

My worry is that the 'forkandmonitor' function is quite complex. If there are blocks that won't be active for the invocation from non-root.go, it will be good for the maintainability for the function to add a few comments.

Reply:
`ForkAndMonitor` was just moved from `virt-launcher.go` to `fork.go`. I would prefer to not touch this part of the code in this PR.